[SNAP-1084] Fix GemFire/Snappy activation classes to return proper classLoaderVersion

sumwale opened the pull request:
## Changes proposed in this pull request

Fix for performance issue seen with queries in cluster mode where simple average on ~100M takes >300ms in single server cluster vs ~150ms in local[*] mode.
- added proper getClassLoaderVersion() impls for GemFireActivationClass and SnappyActivationClass
- cache default TimeZone and Locale statically since those showed up in some perf runs (esp from SnappyData side)
- renamed TraceExecute variable to TraceExecution to match the actual name and avoid confusion
## Patch testing

store precheckin
## ReleaseNotes changes

NA
## Other PRs

SnappyData layer will make use of above changes.

Commit:
[SNAP-1084] Fix GemFireActivationClass and SnappyActivationClass to return proper classLoaderVersion

- added proper getClassLoaderVersion() impls for GemFireActivationClass and SnappyActivationClass
- cache default TimeZone and Locale statically since those showed up in some perf runs (esp from SnappyData side)
- renamed TraceExecute variable to TraceExecution to match the actual name and avoid confusion

soubhik-c reviewed (approved):
Awesome, this was a clear oversight right from day 1 I suppose. TPMG probably is going to rolling out snappy version 0.6, should this fix be put in there and republish binaries as 0.61 ?

Commit:
Added direct conversion to UTF8String and Date/Timestamp

- new methods to directly get Spark's UTF8String from underlying UTF8 bytes
  (still buggy for the case of surrogate pairs where full UTF8 is different from modified UTF8)
- new methods to directly get primitive long values for Date/Timestamp fields
  with values as can be interpreted by Spark (millis for Date and micros for Timestamp)

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/reflect/SnappyActivationClass.java:
```diff
@@ -25,13 +25,19 @@
 
   public int getClassLoaderVersion() {
-    return 0;
+    return this.classLoaderVersion;
```
BTW if instead of 0, -1 would have also led to no recompilation, except where in new jars are being installed in the system. Right? 

Looking at the code it looks like every single time re-compile was happening for statement matching at least and this could be the reason why after switching off statement matching recently in one of the test the numbers were a bit faster. Right?

sumwale replied:
> BTW if instead of 0, -1 would have also led to no recompilation, except where in new jars are being installed in the system. Right? 

I don't think so. The code GenericLanguageConnectionContext compares against current ClassLoaderVersion returned by ClassFactory so -1 won't work (https://github.com/SnappyDataInc/snappy-store/blob/snappy/master/gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/sql/conn/GenericLanguageConnectionContext.java#L1194)

> switching off statement matching recently in one of the test the numbers were a bit faster. Right?

That looks to be a different issue because this recompile will happen even for prepared statements. Normal statements without statement matching will be compiled every time in any case.
